crates/ra_hir_ty/src/infer.rs

```diff
@@ -451,11 +458,15 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             self.infer_pat(*pat, &ty, BindingMode::default());
         }
         let return_ty = self.make_ty(&data.ret_type);
-        self.return_ty = self.insert_vars_for_impl_trait(return_ty);
+        let return_ty = self.insert_vars_for_impl_trait(return_ty);
+        self.return_tys.push(return_ty);
     }
 
     fn infer_body(&mut self) {
-        self.infer_expr(self.body.body_expr, &Expectation::has_type(self.return_ty.clone()));
+        self.infer_expr(
+            self.body.body_expr,
+            &Expectation::has_type(self.return_tys.last().cloned().unwrap_or(Ty::Unknown)),
```

Comment:
```suggestion
            &Expectation::has_type(self.return_tys.last().unwrap().clone()),
```

I think we should guarantee that the stack has at least one element?